bootstrap/kubeadm/api/v1alpha2/conversion.go

```diff
@@ -40,10 +40,23 @@ func (src *KubeadmConfig) ConvertTo(dstRaw conversion.Hub) error {
 	dst.Spec.Verbosity = restored.Spec.Verbosity
 	dst.Spec.UseExperimentalRetryJoin = restored.Spec.UseExperimentalRetryJoin
 	dst.Spec.Files = restored.Spec.Files
-	for i := range restored.Spec.Files {
-		file := restored.Spec.Files[i]
-		if file.ContentFrom != nil {
-			dst.Spec.Files = append(dst.Spec.Files, file)
+
+	// Track files successfully up-converted. We need this to dedupe
+	// restored files from user-updated files on up-conversion. We store
+	// them as pointers for later modification without paying for second
+	// lookup.
+	dstPaths := make(map[string]*kubeadmbootstrapv1alpha3.File, len(dst.Spec.Files))
+	for i := range dst.Spec.Files {
+		file := dst.Spec.Files[i]
+		dstPaths[file.Path] = &file
+	}
+
+	// If we find a restored file matching the file path of a v1alpha2
+	// file with no content, we should restore contentFrom to that file.
+	for _, restoredFile := range restored.Spec.Files {
+		dstFile, exists := dstPaths[restoredFile.Path]
+		if exists && dstFile.Content == "" {
```

Comment:
I read the other thread but I'm still not sure I follow why we're checking for empty Content here instead of checking for set ContentFrom... isn't empty string content a valid File use case?

Reply:
So if you set a file content to empty in either v1a2 or v1a3, that still works. This will only try to match files where a v1a3 file was downconverted into empty content and then re-unconverted. It requires a match on the path.

Realistically, I don’t think this makes any difference. When does downconversion even happen in this scenario? V1a3 is the storage version, this is mostly for the sake of maintaining roundtrip-ability?

Reply:
That said I know string pointer is more correct and this feels weird, but I think it works okay in practice